File: CS/CustomHandlers/AppointmentsDragHelper.cs
using DevExpress.Xpf.Scheduler;
using DevExpress.XtraScheduler;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Windows.Input;
using System.Windows;
using DevExpress.Xpf.Core;

public class AppointmentDragHelper : SchedulerMoveEventHandler {
    AppointmentDragInfo dragInfo;
    DateTime prevDate;

    public DateTime newAppointmentDate { get; set; }
    public bool IsExternalDrag { get; set; }

    protected DateTime CurrentDateTime {
        get { return GetCurrentDateTime(); }
    }

    public AppointmentDragHelper(SchedulerControl control)
        : base(control) {
    }

    public override void AttachToControl() {
        control.MouseDown += MouseDownHandler;
        control.AppointmentDrop += AppointmentDropHandler;
        control.AppointmentDrag += AppointmentDragHandler;
    }

    void control_MouseEnter(object sender, System.Windows.Input.MouseEventArgs e) {

    }

    public override void DetachFromControl() {
        control.MouseDown -= MouseDownHandler;
        control.AppointmentDrop -= AppointmentDropHandler;
        control.AppointmentDrag -= AppointmentDragHandler;
    }

    void MouseDownHandler(object sender, MouseButtonEventArgs e) {
        CreateAppoointmentDragInfo();
    }

    void CreateAppoointmentDragInfo() {
        if(dragInfo == null) {
            if(AppointmentViewInfo != null) {
                IsExternalDrag = false;
                dragInfo = new AppointmentDragInfo() {
                    Appointment = AppointmentViewInfo.Appointment,
                    TimeAtCursor = CurrentDateTime.TimeOfDay,
                    InitialInterval = new TimeInterval(AppointmentViewInfo.Appointment.Start, AppointmentViewInfo.Appointment.End)
                };
            }
            else if(SelectableIntervalViewInfo != null) {
                IsExternalDrag = true;
                dragInfo = new AppointmentDragInfo() {
                    Appointment = null,
                    TimeAtCursor = CurrentDateTime.TimeOfDay,
                    InitialInterval = SelectableIntervalViewInfo.Interval
                };
            }            
        }
    }


    void AppointmentDragHandler(object sender, AppointmentDragEventArgs e) {
        CreateAppoointmentDragInfo();
        OnDragDrop(e);
    }
    void AppointmentDropHandler(object sender, AppointmentDragEventArgs e) {
        OnDragDrop(e);
        dragInfo = null;
    }
    DateTime GetCurrentDateTime() {
        HitTimeCellInfo hitTimeCellInfo = HitTimeCellInfo;
        if(hitTimeCellInfo == null)
            return prevDate;
        prevDate = hitTimeCellInfo.ViewInfo.Interval.Start + hitTimeCellInfo.TimeShift;
        return prevDate;
    }

    void OnDragDrop(AppointmentDragEventArgs e) {
        if(dragInfo == null)
            return;
        if(dragInfo.Appointment == null)
            e.EditedAppointment.Start = GetCurrentDateTime() + dragInfo.InitialInterval.Start.TimeOfDay - dragInfo.TimeAtCursor;
        else
            e.EditedAppointment.Start = GetCurrentDateTime() + e.SourceAppointment.Start.TimeOfDay - dragInfo.TimeAtCursor;

        newAppointmentDate = e.EditedAppointment.Start;
    }
}

File: VB/CustomHandlers/AppointmentsDragHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Xpf.Scheduler
Imports DevExpress.XtraScheduler
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Windows.Input
Imports System.Windows
Imports DevExpress.Xpf.Core

Public Class AppointmentDragHelper
	Inherits SchedulerMoveEventHandler
	Private dragInfo As AppointmentDragInfo
	Private prevDate As DateTime

	Public Property newAppointmentDate() As DateTime
	Public Property IsExternalDrag() As Boolean

	Protected ReadOnly Property CurrentDateTime() As DateTime
		Get
			Return GetCurrentDateTime()
		End Get
	End Property

	Public Sub New(ByVal control As SchedulerControl)
		MyBase.New(control)
	End Sub

	Public Overrides Sub AttachToControl()
		AddHandler control.MouseDown, AddressOf MouseDownHandler
		AddHandler control.AppointmentDrop, AddressOf AppointmentDropHandler
		AddHandler control.AppointmentDrag, AddressOf AppointmentDragHandler
	End Sub

	Private Sub control_MouseEnter(ByVal sender As Object, ByVal e As System.Windows.Input.MouseEventArgs)

	End Sub

	Public Overrides Sub DetachFromControl()
		RemoveHandler control.MouseDown, AddressOf MouseDownHandler
		RemoveHandler control.AppointmentDrop, AddressOf AppointmentDropHandler
		RemoveHandler control.AppointmentDrag, AddressOf AppointmentDragHandler
	End Sub

	Private Sub MouseDownHandler(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
		CreateAppoointmentDragInfo()
	End Sub

	Private Sub CreateAppoointmentDragInfo()
		If dragInfo Is Nothing Then
			If AppointmentViewInfo IsNot Nothing Then
				IsExternalDrag = False
				dragInfo = New AppointmentDragInfo() With {.Appointment = AppointmentViewInfo.Appointment, .TimeAtCursor = CurrentDateTime.TimeOfDay, .InitialInterval = New TimeInterval(AppointmentViewInfo.Appointment.Start, AppointmentViewInfo.Appointment.End)}
			ElseIf SelectableIntervalViewInfo IsNot Nothing Then
				IsExternalDrag = True
				dragInfo = New AppointmentDragInfo() With {.Appointment = Nothing, .TimeAtCursor = CurrentDateTime.TimeOfDay, .InitialInterval = SelectableIntervalViewInfo.Interval}
			End If
		End If
	End Sub


	Private Sub AppointmentDragHandler(ByVal sender As Object, ByVal e As AppointmentDragEventArgs)
		CreateAppoointmentDragInfo()
		OnDragDrop(e)
	End Sub
	Private Sub AppointmentDropHandler(ByVal sender As Object, ByVal e As AppointmentDragEventArgs)
		OnDragDrop(e)
		dragInfo = Nothing
	End Sub
	Private Function GetCurrentDateTime() As DateTime
        Dim hitTimeCellInfo As HitTimeCellInfo = Me.HitTimeCellInfo
        If hitTimeCellInfo Is Nothing Then
            Return prevDate
        End If
		prevDate = hitTimeCellInfo.ViewInfo.Interval.Start + hitTimeCellInfo.TimeShift
		Return prevDate
	End Function

	Private Sub OnDragDrop(ByVal e As AppointmentDragEventArgs)
		If dragInfo Is Nothing Then
			Return
		End If
		If dragInfo.Appointment Is Nothing Then
			e.EditedAppointment.Start = GetCurrentDateTime() + dragInfo.InitialInterval.Start.TimeOfDay - dragInfo.TimeAtCursor
		Else
			e.EditedAppointment.Start = GetCurrentDateTime() + e.SourceAppointment.Start.TimeOfDay - dragInfo.TimeAtCursor
		End If

		newAppointmentDate = e.EditedAppointment.Start
	End Sub
End Class
